File: SignalServiceKit/src/Security/OWSFingerprintBuilder.m
//
// Copyright 2017 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "OWSFingerprintBuilder.h"
#import "ContactsManagerProtocol.h"
#import "OWSFingerprint.h"
#import "OWSIdentityManager.h"
#import "TSAccountManager.h"
#import <Curve25519Kit/Curve25519.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

@interface OWSFingerprintBuilder ()

@property (nonatomic, readonly) TSAccountManager *accountManager;
@property (nonatomic, readonly) id<ContactsManagerProtocol> contactsManager;

@end

@implementation OWSFingerprintBuilder

- (instancetype)initWithAccountManager:(TSAccountManager *)accountManager
                       contactsManager:(id<ContactsManagerProtocol>)contactsManager
{
    self = [super init];
    if (!self) {
        return self;
    }

    _accountManager = accountManager;
    _contactsManager = contactsManager;

    return self;
}

- (nullable OWSFingerprint *)fingerprintWithTheirSignalAddress:(SignalServiceAddress *)theirSignalAddress
{
    NSData *_Nullable theirIdentityKey = [[OWSIdentityManager shared] identityKeyForAddress:theirSignalAddress];

    if (theirIdentityKey == nil) {
        OWSFailDebug(@"Missing their identity key");
        return nil;
    }

    return [self fingerprintWithTheirSignalAddress:theirSignalAddress theirIdentityKey:theirIdentityKey];
}

- (OWSFingerprint *)fingerprintWithTheirSignalAddress:(SignalServiceAddress *)theirSignalAddress
                                     theirIdentityKey:(NSData *)theirIdentityKey
{
    NSString *theirName = [self.contactsManager displayNameForAddress:theirSignalAddress];

    SignalServiceAddress *mySignalAddress = [self.accountManager localAddress];
    // PNI TODO: This should use the identity key associated with our PNI if we only have a PNI session with them.
    NSData *myIdentityKey = [[OWSIdentityManager shared] identityKeyPairForIdentity:OWSIdentityACI].publicKey;

    return [OWSFingerprint fingerprintWithMyStableAddress:mySignalAddress
                                            myIdentityKey:myIdentityKey
                                       theirStableAddress:theirSignalAddress
                                         theirIdentityKey:theirIdentityKey
                                                theirName:theirName];
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/Security/OWSFingerprintBuilder.swift
//
// Copyright 2023 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

public import LibSignalClient

final public class OWSFingerprintBuilder {
    public struct FingerprintResult {
        public let theirAci: Aci
        public let theirRecipientIdentity: OWSRecipientIdentity
        public let fingerprint: OWSFingerprint
    }

    private let contactsManager: any ContactManager
    private let identityManager: OWSIdentityManager
    private let tsAccountManager: TSAccountManager

    public init(
        contactsManager: any ContactManager,
        identityManager: OWSIdentityManager,
        tsAccountManager: TSAccountManager
    ) {
        self.contactsManager = contactsManager
        self.identityManager = identityManager
        self.tsAccountManager = tsAccountManager
    }

    /// Builds fingerprints combining your current credentials with a specified
    /// identity key. You can use these to present a new identity key for
    /// verification.
    public func fingerprints(
        theirAci: Aci,
        theirRecipientIdentity: OWSRecipientIdentity,
        tx: DBReadTransaction
    ) -> FingerprintResult? {
        guard
            let localIdentifiers = tsAccountManager.localIdentifiers(tx: tx),
            let myAciIdentityKey = identityManager.identityKeyPair(for: .aci, tx: tx)?.keyPair.identityKey,
            let theirAciIdentityKey = try? theirRecipientIdentity.identityKeyObject
        else {
            owsFailDebug("Missing local properties!")
            return nil
        }
        let myAci = localIdentifiers.aci

        let theirName = contactsManager.displayName(for: SignalServiceAddress(theirAci), tx: tx).resolvedValue()

        let aciFingerprint = OWSFingerprint(
            myAci: myAci,
            theirAci: theirAci,
            myAciIdentityKey: myAciIdentityKey,
            theirAciIdentityKey: theirAciIdentityKey,
            theirName: theirName
        )

        return FingerprintResult(
            theirAci: theirAci,
            theirRecipientIdentity: theirRecipientIdentity,
            fingerprint: aciFingerprint
        )
    }
}
